flight/flight-core/src/main/java/org/apache/arrow/flight/grpc/NettyClientBuilder.java

```diff
@@ -0,0 +1,232 @@
+   * tracing.
+   */
+  private static final int MAX_CHANNEL_TRACE_EVENTS = 0;
+
+  protected BufferAllocator allocator;
+  protected Location location;
+  protected boolean forceTls = false;
+  protected int maxInboundMessageSize = Integer.MAX_VALUE;
+  protected InputStream trustedCertificates = null;
+  protected InputStream clientCertificate = null;
+  protected InputStream clientKey = null;
+  protected String overrideHostname = null;
+  protected List<FlightClientMiddleware.Factory> middleware = new ArrayList<>();
+  protected boolean verifyServer = true;
+
+  public NettyClientBuilder() {}
+
+  public NettyClientBuilder(BufferAllocator allocator, Location location) {
+    this.allocator = Preconditions.checkNotNull(allocator);
+    this.location = Preconditions.checkNotNull(location);
+  }
+
+  /** Force the client to connect over TLS. */
+  public NettyClientBuilder useTls() {
+    this.forceTls = true;
+    return this;
+  }
+
+  /** Override the hostname checked for TLS. Use with caution in production. */
+  public NettyClientBuilder overrideHostname(final String hostname) {
+    this.overrideHostname = hostname;
+    return this;
+  }
+
+  /** Set the maximum inbound message size. */
+  public NettyClientBuilder maxInboundMessageSize(int maxSize) {
+    Preconditions.checkArgument(maxSize > 0);
+    this.maxInboundMessageSize = maxSize;
+    return this;
+  }
+
+  /** Set the trusted TLS certificates. */
+  public NettyClientBuilder trustedCertificates(final InputStream stream) {
+    this.trustedCertificates = Preconditions.checkNotNull(stream);
+    return this;
+  }
+
+  /** Set the trusted TLS certificates. */
+  public NettyClientBuilder clientCertificate(
+      final InputStream clientCertificate, final InputStream clientKey) {
+    Preconditions.checkNotNull(clientKey);
+    this.clientCertificate = Preconditions.checkNotNull(clientCertificate);
+    this.clientKey = Preconditions.checkNotNull(clientKey);
+    return this;
+  }
+
+  public BufferAllocator allocator() {
+    return allocator;
+  }
+
+  public NettyClientBuilder allocator(BufferAllocator allocator) {
+    this.allocator = Preconditions.checkNotNull(allocator);
+    return this;
+  }
+
+  public NettyClientBuilder location(Location location) {
+    this.location = Preconditions.checkNotNull(location);
+    return this;
+  }
+
+  public List<FlightClientMiddleware.Factory> middleware() {
+    return Collections.unmodifiableList(middleware);
+  }
+
+  public NettyClientBuilder intercept(FlightClientMiddleware.Factory factory) {
+    middleware.add(factory);
+    return this;
+  }
+
+  public NettyClientBuilder verifyServer(boolean verifyServer) {
+    this.verifyServer = verifyServer;
+    return this;
+  }
+
+  /** Create the client from this builder. */
+  public NettyChannelBuilder build() {
+    final NettyChannelBuilder builder;
+
+    switch (location.getUri().getScheme()) {
```

Comment:
FYI, I'm investigating the Netty channel address depending of the scheme as it seems we have a weird behavior here.
Your change move that from `FlightClient` to `NettyClientBuilder` so it looks good to me.